lib/xnvme_be_linux_async_liburing.c

```diff
@@ -392,6 +402,37 @@ xnvme_be_linux_liburing_cmd_iov(struct xnvme_cmd_ctx *ctx, struct iovec *dvec, s
 
 	return 0;
 }
+
+int
+xnvme_be_linux_liburing_get_completion_fd(struct xnvme_queue *queue)
+{
+	struct xnvme_queue_liburing *q = (struct xnvme_queue_liburing *)queue;
+	int efd;
+
+	if (q->efd != -1) {
+		return q->efd;
+	}
+
+	efd = eventfd(0, EFD_CLOEXEC);
+	if (efd < 0) {
+		XNVME_DEBUG("FAILED: failed to create eventfd");
+		return -errno;
+	}
+
+	q->efd = efd;
+
+	if (io_uring_register_eventfd(&q->ring, efd)) {
+		XNVME_DEBUG("FAILED: failed to register eventfd");
+		close(efd);
+		return -errno;
+	}
+
+	q->batching = 0;
```

Comment:
If the user has submitted one or more commands without calling ``poke()``, there will be unsubmitted commands left in the queue. To prevent this, consider making this part more defensive.

For example, it could check for outstanding I/O and either submit the commands or return an error like ``EBUSY`` to indicate the queue must be cleared before calling ``get_completion_fd()``. I would lean towards the latter.

Reply:
The outstanding commands would be submitted on the next call to `...cmd_io()`/`...cmd_iov()`.
Is that not good enough?
To me, returning `-EBUSY` when trying to get a `completion_fd` doesn't map intuitively to "there are outstanding commands".

Reply:
They would, if it were ever invoked. However, it is likely that someone might jump into the API and perform a code experiment, such as invoking ``cmd_io()`` one or more times, then calling ``xnvme_queue_get_completion_fd()`` to retrieve an eventfd to start waiting for those command submissions to complete. This would simply result in a hang, with no information provided.

To avoid this, I recommend using the error value, which is also utilized to indicate: "the queue is doing something else, so you cannot set up or retrieve an eventfd at this time." By adding this information to the docstring of ``xnvme_queue_get_completion_fd()``, users would be guided toward the correct approach.